### ISSUE TITLE
Java8 Windows has incorrect vcruntime140.dll name, vcrunt~.dll

### ISSUE THREAD

**Issue by NagarajuGunda** [eclipse-openj9/openj9#15962]
Java -version output
--------------------

Output from `java -version`.
c:\>C:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\java -version
jvm.dll preloadLibrary: LoadLibrary(C:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\default\j9vm29.dll) error: 7e
jvm.dll preloadLibrary: LoadLibrary(C:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\default\j9hookable29.dll) error: 7e
jvm.dll failed to load: j9vm29

From the bin directory

C:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin>java -version
openjdk version "1.8.0_345"
IBM Semeru Runtime Open Edition (build 1.8.0_345-b01)
Eclipse OpenJ9 VM (build openj9-0.33.1, JRE 1.8.0 Windows Server 2022 amd64-64-Bit Compressed References 20220817_464 (JIT enabled, AOT enabled)
OpenJ9   - 1d9d16830
OMR      - b58aa2708
JCL      - e361c66299 based on jdk8u345-b01)

Summary of problem
------------------

Java 1.8.0_345-b01 is not working with full path from the directory other than bin.

**Comment by hangshao0** [comment]
It is working for me on my Windows 10:

```
C:\>C:\Users\HANGShao\Downloads\ibm-semeru-open-jre_x64_windows_8u345b01_openj9-0.33.1\jdk8u345-b01-jre\bin\java -version
openjdk version "1.8.0_345"
IBM Semeru Runtime Open Edition (build 1.8.0_345-b01)
Eclipse OpenJ9 VM (build openj9-0.33.1, JRE 1.8.0 Windows 10 amd64-64-Bit Compressed References 20220817_464 (JIT enabled, AOT enabled)
OpenJ9   - 1d9d16830
OMR      - b58aa2708
JCL      - e361c66299 based on jdk8u345-b01)
```

Does it fail on earlier releases on your side ?

**Comment by NagarajuGunda** [comment]
> This seems like a dll dependency problem. It may depend on whether dependencies of the dlls are generally available on the machine, in particular the Universal C Runtime to support Visual Studio 2017. Although they are available in the bin directory, it seems like they are not being found there. I'm guessing if the Universal C Runtime is installed on the machine, the problem won't be seen.

I am trying this on a new machine where no runtime or vcredist is installed. Reason being, we use this Java within our installer (InstallAnywhere) and it used java to launch the installer. If the java doesn't work, then the installer itself wouldn't launch.

I am observing this issue only with recent JRE builds with VS 2017 as the dependent dll msvcp140.dll is in turn dependent on vcruntime140.dll which was not the case earlier. Earlier, java dll's were only dependent on msvcp120.dll and vcruntime*.dll was not present.

There is another problem here. vcruntime140.dll is shipped as vcrunt~.dll and I had to rename it.

After going through preloadLibrary code, we figured it might be a problem with the way the preload libraries are being Loaded. 

https://github.com/eclipse-openj9/openj9/blob/v0.27.0-release/runtime/j9vm/jvm.c#:~:text=LOAD_WITH_ALTERED_SEARCH_PATH

Since, altered searched strategy is used instead of standard search strategy, vcruntime140.dll is being searched in dll directory instead of application directories for certain LoadLibrary calls.

Please validate if my understanding is correct :)

**Comment by NagarajuGunda** [comment]
> It is working for me on my Windows 10:
> 
> ```
> C:\>C:\Users\HANGShao\Downloads\ibm-semeru-open-jre_x64_windows_8u345b01_openj9-0.33.1\jdk8u345-b01-jre\bin\java -version
> openjdk version "1.8.0_345"
> IBM Semeru Runtime Open Edition (build 1.8.0_345-b01)
> Eclipse OpenJ9 VM (build openj9-0.33.1, JRE 1.8.0 Windows 10 amd64-64-Bit Compressed References 20220817_464 (JIT enabled, AOT enabled)
> OpenJ9   - 1d9d16830
> OMR      - b58aa2708
> JCL      - e361c66299 based on jdk8u345-b01)
> ```
> 
> Does it fail on earlier releases on your side ?



> 1.8.0_

1.8.0_332 worked while 1.8.0_345 did not

c:\>c:\Users\Administrator\Downloads\jdk8u332-b09-jre\bin\java -version
openjdk version "1.8.0_332"
IBM Semeru Runtime Open Edition (build 1.8.0_332-b09)
Eclipse OpenJ9 VM (build openj9-0.32.0, JRE 1.8.0 Windows Server 2022 amd64-64-Bit Compressed References 20220422_375 (JIT enabled, AOT enabled)
OpenJ9   - 9a84ec34e
OMR      - ab24b6666
JCL      - 0b8b8af39a based on jdk8u332-b09)

c:\>c:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\java -version
jvm.dll preloadLibrary: LoadLibrary(c:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\default\j9vm29.dll) error: 7e
jvm.dll preloadLibrary: LoadLibrary(c:\Users\Administrator\Downloads\jdk8u345-b01-jre\bin\default\j9hookable29.dll) error: 7e
jvm.dll failed to load: j9vm29

I believe you can reproduce the issue by removing the installed vc redists

**Comment by pshipton** [comment]
The difference between 332 and 345 is the compiler changed from VS2013 to VS2017, adding the requirement for the Universal C Runtime.

**Comment by pshipton** [comment]
We'll have to figure out what is actually going wrong. There is code run from java.exe to pre-load vcruntime140.dll before jvm.dll is loaded, and it seems to be working for me. 

On a Windows 10 machine, this is what I see when running with the debugger (after fixing the vcruntime140.dll name). You can see that vcruntime140.dll and msvcp140.dll are loaded first from the bin directory.

```
Microsoft (R) Windows Debugger Version 6.12.0002.633 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

CommandLine: j2sdk-image\bin\java -version
Symbol search path is: C:\Users\PETERShipton\Downloads
Executable search path is: C:\Users\PETERShipton\Downloads
ModLoad: 00007ff7`db0d0000 00007ff7`db115000   java.exe
ModLoad: 00007ffd`c5950000 00007ffd`c5b48000   ntdll.dll
ModLoad: 00007ffd`c48e0000 00007ffd`c499d000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007ffd`c3640000 00007ffd`c390e000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007ffd`c4490000 00007ffd`c453e000   C:\WINDOWS\System32\ADVAPI32.dll
ModLoad: 00007ffd`c43f0000 00007ffd`c448e000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007ffd`c4ee0000 00007ffd`c4f7c000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007ffd`c4690000 00007ffd`c47b5000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007ffd`c3de0000 00007ffd`c3f80000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007ffd`c34f0000 00007ffd`c3512000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007ffd`c4eb0000 00007ffd`c4eda000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007ffd`c3150000 00007ffd`c325b000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007ffd`c3910000 00007ffd`c39ad000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007ffd`c3260000 00007ffd`c3360000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007ffd`afb50000 00007ffd`afdea000   C:\WINDOWS\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\COMCTL32.dll
(1264.1a28): Break instruction exception - code 80000003 (first chance)
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for ntdll.dll - 
ntdll!LdrInitShimEngineDynamic+0x360:
00007ffd`c5a20930 cc              int     3
0:000> g
ModLoad: 00007ffd`c48a0000 00007ffd`c48d0000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00007ffd`be980000 00007ffd`be995000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\vcruntime140.dll
ModLoad: 00007ffd`be570000 00007ffd`be60b000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\msvcp140.dll
ModLoad: 00007ffd`be520000 00007ffd`be538000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\j9vm\jvm.dll
ModLoad: 00007ffd`c17c0000 00007ffd`c17d2000   C:\WINDOWS\SYSTEM32\kernel.appcore.dll
ModLoad: 00007ffd`bb940000 00007ffd`bb988000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\jvm.dll
ModLoad: 00007ffd`c49a0000 00007ffd`c4a0b000   C:\WINDOWS\System32\WS2_32.dll
ModLoad: 00007ffd`9f950000 00007ffd`9fb4e000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9vm29.dll
ModLoad: 00007ffd`bd410000 00007ffd`bd427000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9thr29.dll
ModLoad: 00007ffd`be970000 00007ffd`be97b000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9hookable29.dll
ModLoad: 00007ffd`bb640000 00007ffd`bb6aa000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9prt29.dll
ModLoad: 00007ffd`c51c0000 00007ffd`c5903000   C:\WINDOWS\System32\SHELL32.dll
ModLoad: 00007ffd`be8d0000 00007ffd`be8d8000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\omrsig.dll
ModLoad: 00007ffd`a0ff0000 00007ffd`a1039000   C:\WINDOWS\SYSTEM32\pdh.dll
ModLoad: 00007ffd`bcdd0000 00007ffd`bcde8000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9zlib29.dll
ModLoad: 00007ffd`c30c0000 00007ffd`c3142000   C:\WINDOWS\System32\bcryptPrimitives.dll
ModLoad: 00007ffd`bb8f0000 00007ffd`bb939000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9dmp29.dll
ModLoad: 00007ffd`bb600000 00007ffd`bb638000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9trc29.dll
ModLoad: 00007ffd`c2f30000 00007ffd`c2f62000   C:\WINDOWS\SYSTEM32\SspiCli.dll
ModLoad: 00007ffd`6cfc0000 00007ffd`6d8ac000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9jit29.dll
ModLoad: 00007ffd`9efc0000 00007ffd`9f393000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9gc29.dll
ModLoad: 00007ffd`a4410000 00007ffd`a44df000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9vrb29.dll
ModLoad: 00007ffd`a57c0000 00007ffd`a586d000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9shr29.dll
ModLoad: 00007ffd`a5320000 00007ffd`a5380000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\j9jvmti29.dll
ModLoad: 00007ffd`a2f10000 00007ffd`a2f96000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\default\jclse29.dll
ModLoad: 00007ffd`b1d80000 00007ffd`b1d97000   C:\WINDOWS\system32\napinsp.dll
ModLoad: 00007ffd`b1d20000 00007ffd`b1d3b000   C:\WINDOWS\system32\pnrpnsp.dll
ModLoad: 00007ffd`bd9c0000 00007ffd`bd9d5000   C:\WINDOWS\system32\wshbth.dll
ModLoad: 00007ffd`bb4e0000 00007ffd`bb4fd000   C:\WINDOWS\system32\NLAapi.dll
ModLoad: 00007ffd`c24d0000 00007ffd`c250b000   C:\WINDOWS\SYSTEM32\IPHLPAPI.DLL
ModLoad: 00007ffd`c2830000 00007ffd`c289a000   C:\WINDOWS\System32\mswsock.dll
ModLoad: 00007ffd`c2520000 00007ffd`c25ea000   C:\WINDOWS\SYSTEM32\DNSAPI.dll
ModLoad: 00007ffd`c5090000 00007ffd`c5098000   C:\WINDOWS\System32\NSI.dll
ModLoad: 00007ffd`b1b30000 00007ffd`b1b42000   C:\WINDOWS\System32\winrnr.dll
ModLoad: 00007ffd`b8650000 00007ffd`b865a000   C:\Windows\System32\rasadhlp.dll
ModLoad: 00007ffd`b8090000 00007ffd`b8110000   C:\WINDOWS\System32\fwpuclnt.dll
ModLoad: 00007ffd`c34c0000 00007ffd`c34e7000   C:\WINDOWS\System32\bcrypt.dll
ModLoad: 00007ffd`baa00000 00007ffd`baa0f000   C:\WINDOWS\System32\perfos.dll
ModLoad: 00007ffd`c0fe0000 00007ffd`c1775000   C:\WINDOWS\SYSTEM32\windows.storage.dll
ModLoad: 00007ffd`c3a80000 00007ffd`c3dd4000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007ffd`c2ad0000 00007ffd`c2b00000   C:\WINDOWS\System32\Wldp.dll
ModLoad: 00007ffd`c4e00000 00007ffd`c4ead000   C:\WINDOWS\System32\SHCORE.dll
ModLoad: 00007ffd`c47c0000 00007ffd`c4815000   C:\WINDOWS\System32\shlwapi.dll
ModLoad: 00007ffd`ba620000 00007ffd`ba643000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\java.dll
ModLoad: 00007ffd`ba620000 00007ffd`ba643000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\java.dll
ModLoad: 00007ffd`be560000 00007ffd`be570000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\verify.dll
ModLoad: 00007ffd`bcbb0000 00007ffd`bcbc8000   C:\Users\PETERShipton\Downloads\j2sdk-image\jre\bin\zip.dll
(1264.404): C++ EH exception - code e06d7363 (first chance)
(1264.404): C++ EH exception - code e06d7363 (first chance)
ntdll!NtTerminateProcess+0x14:
00007ffd`c59ed584 c3              ret
```

**Comment by pshipton** [comment]
I think fixing the name of vcruntime140.dll in the build fixes the problem. Without the proper name, the pre-loading code doesn't work. i.e. https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/615

There is a build linked below you can try to confirm.

I created a Windows 2022 machine and duplicated the problem. The fixed build below is working fine for me on this machine.

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Build_JDK8_x86-64_windows_Personal/519/OpenJ9-JDK8-x86-64_windows-20220927-141014.tar.gz

**Comment by NagarajuGunda** [comment]
Hi @pshipton 

The build you have provided seems to be working fine on the same machine. I am just curious why it wouldn't have worked earlier even though I renamed the vcrunt~.dll to vcruntime140.dll

'''
c:\>c:\Users\Administrator\Downloads\j2sdk-image\jre\bin\java -version
openjdk version "1.8.0_352-internal"
OpenJDK Runtime Environment (build 1.8.0_352-internal-jenkins_2022_09_27_14_10-b00)
Eclipse OpenJ9 VM (build master-4e055dc9920, JRE 1.8.0 Windows Server 2022 amd64-64-Bit Compressed References 20220927_519 (JIT enabled, AOT enabled)
OpenJ9   - 4e055dc9920
OMR      - 24b511d
JCL      - cb21f32142e based on jdk8u352-b06)

c:\>c:\Users\Administrator\Downloads\j2sdk-image\bin\java -version
openjdk version "1.8.0_352-internal"
OpenJDK Runtime Environment (build 1.8.0_352-internal-jenkins_2022_09_27_14_10-b00)
Eclipse OpenJ9 VM (build master-4e055dc9920, JRE 1.8.0 Windows Server 2022 amd64-64-Bit Compressed References 20220927_519 (JIT enabled, AOT enabled)
OpenJ9   - 4e055dc9920
OMR      - 24b511d
JCL      - cb21f32142e based on jdk8u352-b06)
'''

**Comment by NagarajuGunda** [comment]
After copying vcruntime140.dll to jdk\bin and jdk\jre\bin directories, jdk\bin\java worked but jre\bin\java still has problem (1.8.0_345)

```
c:\>C:\Users\Administrator\Downloads\jdk8u345-b01\bin\java -version
Error: loading: C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\msvcp140.dll
Error: loading: C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\j9vm\jvm.dll

c:\>del "C:\Users\Administrator\Downloads\jdk8u345-b01\bin\vcrunt~1.dll"

c:\>copy C:\Users\Administrator\Downloads\j2sdk-image\bin\vcruntime140.dll C:\Users\Administrator\Downloads\jdk8u345-b01\bin
        1 file(s) copied.

c:\>C:\Users\Administrator\Downloads\jdk8u345-b01\bin\java -version
openjdk version "1.8.0_345"
IBM Semeru Runtime Open Edition (build 1.8.0_345-b01)
Eclipse OpenJ9 VM (build openj9-0.33.1, JRE 1.8.0 Windows Server 2022 amd64-64-Bit Compressed References 20220817_464 (JIT enabled, AOT enabled)
OpenJ9   - 1d9d16830
OMR      - b58aa2708
JCL      - e361c66299 based on jdk8u345-b01)
```

```
c:\>C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\java -version
Error: loading: C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\msvcp140.dll
Error: loading: C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\j9vm\jvm.dll

c:\>del "C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\vcrunt~1.dll"

c:\>copy C:\Users\Administrator\Downloads\j2sdk-image\jre\bin\vcruntime140.dll C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin
        1 file(s) copied.

c:\>C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\java -version
jvm.dll preloadLibrary: LoadLibrary(C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\default\j9vm29.dll) error: 7e
jvm.dll preloadLibrary: LoadLibrary(C:\Users\Administrator\Downloads\jdk8u345-b01\jre\bin\default\j9hookable29.dll) error: 7e
jvm.dll failed to load: j9vm29

```

**Comment by pshipton** [comment]
>  I am just curious why it wouldn't have worked earlier even though I renamed the vcrunt~.dll to vcruntime140.dll

It's not just the file that was renamed, the code in the JVM that pre-loads vcruntime140.dll also changed. Both changes are required for it to work.

Since the test build worked, I'll close this as fixed in the next release, 0.35. There is a Milestone 2 build for 0.35 due out next week that will contain the fix.

**Comment by NagarajuGunda** [comment]
Thanks @pshipton
